File: src/rules/validator.ts
/**
 * Play validation: enumerate all legal plays from a hand given the current trick state.
 *
 * Pure functions — the engine calls these to get legal actions for the current player.
 */

import { Card, Suit, Rank } from '../core/card';
import { PlayType, type PlayAction } from '../core/types';
import { detectSingles, detectPairs, detectTractors, detectThrowCandidates } from './pattern';
import { getLegalFollows } from './follow-rules';

/**
 * Enumerate all legal plays for the current player.
 *
 * @param hand - Current player's hand.
 * @param currentTrick - The current trick slots (null = not yet played).
 * @param trumpSuit - The trump suit.
 * @param trumpRank - The trump rank (= current level).
 * @param isLeading - True if this player is leading the trick.
 * @param leadPlayType - The play type of the lead (null if leading).
 * @param remainingCards - All cards not yet played (for throw validation).
 */
export function getLegalPlays(
  hand: Card[],
  currentTrick: { playerIndex: number; cards: Card[] | null }[],
  trumpSuit: Suit,
  trumpRank: Rank,
  isLeading: boolean,
  leadAction: PlayAction | null,
  remainingCards?: Card[],
): PlayAction[] {
  if (hand.length === 0) return [];

  if (isLeading || currentTrick.every(s => s.cards === null)) {
    return getLeadingPlays(hand, trumpSuit, trumpRank, remainingCards ?? []);
  }

  // Following
  if (!leadAction) return [];

  return getLegalFollows(hand, leadAction, trumpSuit, trumpRank);
}

/**
 * Get all legal plays when leading a trick.
 */
export function getLeadingPlays(
  hand: Card[],
  trumpSuit: Suit,
  trumpRank: Rank,
  remainingCards: Card[],
): PlayAction[] {
  const plays: PlayAction[] = [];

  // Single: every card
  plays.push(...detectSingles(hand));

  // Pair: every identical pair
  plays.push(...detectPairs(hand));

  // Tractor: consecutive pairs
  plays.push(...detectTractors(hand, trumpSuit, trumpRank));

  // Throw: for each non-trump suit
  for (const suit of [Suit.HEARTS, Suit.SPADES, Suit.DIAMONDS, Suit.CLUBS]) {
    if (suit === trumpSuit) continue;
    const candidates = detectThrowCandidates(hand, suit, trumpSuit, trumpRank);
    // Filter throws: each card must be unbeatable by remaining cards of that suit
    for (const candidate of candidates) {
      if (isThrowValid(candidate, suit, trumpRank, remainingCards)) {
        plays.push(candidate);
      }
    }
  }

  return plays;
}

/**
 * Check if a throw is valid: all thrown cards must be the highest remaining
 * cards of that suit among all players.
 */
function isThrowValid(
  throwAction: PlayAction,
  suit: Suit,
  trumpRank: Rank,
  remainingCards: Card[],
): boolean {
  if (remainingCards.length === 0) return true;

  // Find the highest remaining card of this suit (excluding trump rank)
  const remainingOfSuit = remainingCards.filter(
    c => c.suit === suit && c.rank !== trumpRank
  );

  if (remainingOfSuit.length === 0) return true;

  // Sort remaining by rank descending
  const sortedRemaining = [...remainingOfSuit].sort((a, b) => {
    const rankOrder: Record<string, number> = {
      '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8,
      '9': 9, '10': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14,
    };
    return (rankOrder[b.rank] || 0) - (rankOrder[a.rank] || 0);
  });

  const highestRemainingRank = sortedRemaining[0].rank;

  const rankOrder: Record<string, number> = {
    '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8,
    '9': 9, '10': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14,
  };

  // Each thrown card must be >= the highest remaining card of that suit
  return throwAction.cards.every(c =>
    (rankOrder[c.rank] || 0) >= (rankOrder[highestRemainingRank] || 0)
  );
}

/**
 * Filter legal plays to only those matching a specific play type.
 */
export function filterByType(plays: PlayAction[], type: PlayType): PlayAction[] {
  return plays.filter(p => p.type === type);
}

/**
 * Check if a specific play action is legal.
 */
export function isLegalPlay(
  action: PlayAction,
  legalPlays: PlayAction[],
): boolean {
  const actionIds = new Set(action.cards.map(c => c.id));
  return legalPlays.some(legal => {
    const legalIds = new Set(legal.cards.map(c => c.id));
    if (actionIds.size !== legalIds.size) return false;
    for (const id of actionIds) {
      if (!legalIds.has(id)) return false;
    }
    return true;
  });
}

/**
 * Get a human-readable list of legal play descriptions.
 * Useful for AI prompts.
 */
export function describeLegalPlays(plays: PlayAction[]): string[] {
  return plays.map(p => {
    const cards = p.cards.map(c => {
      if (c.isJoker) return c.isBigJoker ? '大王' : '小王';
      const s: Record<string, string> = {
        hearts: '♥', spades: '♠', diamonds: '♦', clubs: '♣',
      };
      return `${s[c.suit]}${c.rank}`;
    }).join(' ');

    switch (p.type) {
      case PlayType.SINGLE: return `单张: ${cards}`;
      case PlayType.PAIR: return `对子: ${cards}`;
      case PlayType.TRACTOR: return `拖拉机(${p.cards.length / 2}对): ${cards}`;
      case PlayType.THROW: return `甩牌: ${cards}`;
    }
  });
}

File: server/rules/validator.py
"""Play validation: enumerate all legal plays from a hand given the current trick state.

Pure functions -- the engine calls these to get legal actions for the current player.

Ported from src/rules/validator.ts.
"""

from __future__ import annotations

from server.engine.card import Card, Suit, Rank
from server.engine.card_utils import RANK_ORDER
from server.engine.types import PlayType, PlayAction
from server.rules.pattern import (
    detect_singles, detect_pairs, detect_tractors, detect_throw_candidates,
    describe_play,
)
from server.rules.follow_rules import get_legal_follows


# ---- Public API ----


def get_legal_plays(
    hand: list[Card],
    current_trick: list[dict],
    trump_suit: Suit,
    trump_rank: Rank,
    is_leading: bool,
    lead_action: PlayAction | None,
    remaining_cards: list[Card] | None = None,
) -> list[PlayAction]:
    """Enumerate all legal plays for the current player.

    Args:
        hand: Current player's hand.
        current_trick: The current trick slots (list of dicts with player_index and cards).
        trump_suit: The trump suit.
        trump_rank: The trump rank (= current level).
        is_leading: True if this player is leading the trick.
        lead_action: The lead play action (None if leading).
        remaining_cards: All cards not yet played (for throw validation).

    Returns:
        List of all legal PlayAction options.
    """
    if not hand:
        return []

    all_slots_empty = all(
        slot.get("cards") is None or slot.get("cards") == []
        for slot in current_trick
    )

    if is_leading or all_slots_empty:
        return get_leading_plays(hand, trump_suit, trump_rank, remaining_cards or [])

    # Following
    if lead_action is None:
        return []

    return get_legal_follows(hand, lead_action, trump_suit, trump_rank)


def get_leading_plays(
    hand: list[Card],
    trump_suit: Suit,
    trump_rank: Rank,
    remaining_cards: list[Card],
) -> list[PlayAction]:
    """Get all legal plays when leading a trick.

    Includes singles, pairs, tractors, and valid throws.
    """
    plays: list[PlayAction] = []

    # Single: every card
    plays.extend(detect_singles(hand))

    # Pair: every identical pair
    plays.extend(detect_pairs(hand))

    # Tractor: consecutive pairs
    plays.extend(detect_tractors(hand, trump_suit, trump_rank))

    # Throw: for each non-trump suit
    for suit in (Suit.HEARTS, Suit.SPADES, Suit.DIAMONDS, Suit.CLUBS):
        if suit == trump_suit:
            continue
        candidates = detect_throw_candidates(hand, suit, trump_suit, trump_rank)
        # Filter throws: each card must be unbeatable by remaining cards of that suit
        for candidate in candidates:
            if _is_throw_valid(candidate, suit, trump_rank, remaining_cards):
                plays.append(candidate)

    return plays


def is_legal_play(action: PlayAction, legal_plays: list[PlayAction]) -> bool:
    """Check if a specific play action is legal.

    Compares by card IDs (set equality).
    """
    action_ids = {c.id for c in action.cards}
    return any(
        action_ids == {c.id for c in legal.cards}
        for legal in legal_plays
    )


def filter_by_type(plays: list[PlayAction], play_type: PlayType) -> list[PlayAction]:
    """Filter legal plays to only those matching a specific play type."""
    return [p for p in plays if p.type == play_type]


def describe_legal_plays(plays: list[PlayAction]) -> list[str]:
    """Get a human-readable list of legal play descriptions.

    Useful for AI prompts.
    """
    return [describe_play(p) for p in plays]


# ---- Private helpers ----


def _is_throw_valid(
    throw_action: PlayAction,
    suit: Suit,
    trump_rank: Rank,
    remaining_cards: list[Card],
) -> bool:
    """Check if a throw is valid: all thrown cards must be the highest remaining
    cards of that suit among all players.
    """
    if not remaining_cards:
        return True

    # Find the highest remaining card of this suit (excluding trump rank)
    remaining_of_suit = [
        c for c in remaining_cards
        if c.suit == suit and c.rank != trump_rank
    ]

    if not remaining_of_suit:
        return True

    # Sort remaining by rank descending
    sorted_remaining = sorted(
        remaining_of_suit,
        key=lambda c: RANK_ORDER[c.rank],
        reverse=True,
    )

    highest_remaining_rank = sorted_remaining[0].rank

    # Each thrown card must be >= the highest remaining card of that suit
    return all(
        RANK_ORDER[c.rank] >= RANK_ORDER[highest_remaining_rank]
        for c in throw_action.cards
    )
